### PR TITLE
Microsoft SQL Server e2e

### PR DESCRIPTION
**Pull request by Vipinofficial11**

### CODE REVIEW

File: mssql-plugin/src/e2e-test/features/mssql/mssql sink/DesignTime.feature [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+## Unless required by applicable law or agreed to in writing, software
+## distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+## WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+## License for the specific language governing permissions and limitations under
+# the License..
+
+@Mssql
+Feature: MySQL Server Sink - Design time scenarios
+
+  @MSSQL_SOURCE_TEST @MSSQL_SINK_TEST @Mssql_Required
```
Comment (Vipinofficial11) [inline mssql-plugin/src/e2e-test/features/mssql/mssql sink/DesignTime.feature]: We don't need @Mssql_Required tag in each scenario

File: mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    And Verify the preview run status of pipeline in the logs is "succeeded"
+    And Close the pipeline logs
+    And Close the preview
+    Then Deploy the pipeline
+    Then Run the Pipeline in Runtime
+    Then Wait till pipeline is in running state
+    Then Open and capture logs
+    Then Verify the pipeline status is "Succeeded"
+    Then Close the pipeline logs
+    And Close the preview
```
Comment (Vipinofficial11) [inline mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature]: These steps are wrong, why are you Deploying the pipeline 2 times. remove unnecessary steps

File: mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+    Then Deploy the pipeline
+    Then Run the Pipeline in Runtime
+    Then Wait till pipeline is in running state
+    Then Open and capture logs
+    Then Verify the pipeline status is "Succeeded"
+    Then Close the pipeline logs
+
+  @MSSQL_SOURCE_TEST @MSSQL_SINK_TEST @Mssql_Required
+  Scenario: To verify data is getting transferred from Mssql to BigQuery successfully
```
Comment (Vipinofficial11) [inline mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature:63]: Already written a step to run a pipeline from MsSql to BigQuery, why we need a second step.

File: mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    Then Close the pipeline logs
+    And Close the preview
+    Then Deploy the pipeline
+    Then Run the Pipeline in Runtime
+    Then Wait till pipeline is in running state
+    Then Open and capture logs
+    Then Verify the pipeline status is "Succeeded"
+    Then Close the pipeline logs
+
+  @MSSQL_SOURCE_TEST @BQ_SINK @Mssql_Required
```
Comment (Vipinofficial11) [inline mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature]: We don't have to add Required tag in each test case. Only add the tag on few tests which you think should run prior to whole build.

File: mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+  Scenario: To verify the pipeline fails while preview with invalid bounding query setting the required split-By field
```
Comment (Vipinofficial11) [inline mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature]: Instead of checking for a failure on Preview, directly Deploy this pipeline and check the error message provided in logs.

File: mssql-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks.java [hunk 6]
```diff
@@ -80,4 +83,11 @@ public static void dropTables() throws SQLException, ClassNotFoundException {
                         PluginPropertyUtils.pluginProp("targetTable")});
     }
 
+    @Before(order = 1, value = "@BQ_SINK")
+    public static void setTempTargetBQTable() {
```
Comment (Vipinofficial11) [inline mssql-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks.java:102]: Write a After Hook as well to delete this target table.

File: mssql-plugin/src/e2e-test/resources/pluginParameters.properties [hunk 7]
```diff
@@ -8,6 +8,26 @@ username=MSSQL_USERNAME
...
+invalidRef=invalidRef&^*&&*
+zeroValue=0
+blankvalue=
+invalidTableName=Table123@
+bqtarget.table=target-table
+project.id=cdf-athena
+datasetprojectId=cdf-athena
+invalid.boundQuery=SELECT MIN(id),MAX(id) FROM table
+splitby=id
+mssqlimportQuery=select * from SOURCETABLE_UUXHVMWPVL where $CONDITIONS;
```
Comment (Vipinofficial11) [inline mssql-plugin/src/e2e-test/resources/pluginParameters.properties]: Don't hardcode the table name here, create a hook in TestSetup Hook and create the query there.  Then use PluginPropertyUtils.addPluginProp() to update this field ``` mssqlimportQuery ```

File: mssql-plugin/src/e2e-test/features/mssql/mssql sink/DesignTime.feature (hunk 1 above)
Comment (priyabhatnagar25) [reply mssql-plugin/src/e2e-test/features/mssql/mssql sink/DesignTime.feature]: resolved

File: mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature (hunk 2 above)
Comment (priyabhatnagar25) [reply mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature]: Resolved

File: mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature (hunk 3 above)
Comment (priyabhatnagar25) [reply mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature:63]: That step was for multiple datatypes and this step is records transfer from mssql to bigquery

File: mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature (hunk 4 above)
Comment (priyabhatnagar25) [reply mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature]: resolved

File: mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature (hunk 5 above)
Comment (priyabhatnagar25) [reply mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature]: resolved

File: mssql-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks.java (hunk 6 above)
Comment (priyabhatnagar25) [reply mssql-plugin/src/e2e-test/java/io.cdap.plugin/common.stepsdesign/TestSetupHooks.java:102]: resolved

File: mssql-plugin/src/e2e-test/features/mssql/mssql sink/RunTimeWithMacros.feature [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+    Then Run the Pipeline in Runtime
+    Then Enter runtime argument value "databaseName" for key "DatabaseName"
+    Then Enter runtime argument value "fetchSize" for key "fetchSize"
+    And Run the Pipeline in Runtime with runtime arguments
+    Then Wait till pipeline is in running state
+    Then Open and capture logs
+    Then Verify the pipeline status is "Succeeded"
+
+  @BQ_SOURCE_TEST @MSSQL_TEST_TABLE @Mssql_Required
+  Scenario: Verify pipeline failure message in logs when user provides invalid Table Name of plugin with Macros
```
Comment (Vipinofficial11) [inline mssql-plugin/src/e2e-test/features/mssql/mssql sink/RunTimeWithMacros.feature:117]: This Step definition doesn't look relatable we are simply checking if pipeline is failing with invalid macro properties and not the failure message.

File: mssql-plugin/src/e2e-test/features/mssql/mssql sink/RunTimeWithMacros.feature [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+    Then Enter input plugin property: "referenceName" with value: "targetRef"
+    Then Validate "SQL Server2" plugin properties
+    Then Save and Deploy Pipeline
+    Then Run the Pipeline in Runtime
+    Then Wait till pipeline is in running state
+    Then Open and capture logs
+    And Verify the pipeline status is "Failed"
+
+  @BQ_SOURCE_TEST @MSSQL_TEST_TABLE @Mssql_Required
+  Scenario: Verify pipeline failure message in logs when user provides invalid Credentials for connection with Macros
```
Comment (Vipinofficial11) [inline mssql-plugin/src/e2e-test/features/mssql/mssql sink/RunTimeWithMacros.feature:160]: Same comment here we are not verifying any message simply just checking if pipeline is failed.

File: mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature (hunk 5 above)
Comment (Vipinofficial11) [reply mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature]: We are still checking if the pipeline is failed not checking the error message.

File: mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTimeWithMacros.feature [hunk 10]
```diff
@@ -115,3 +115,169 @@ Feature: Mssql Source - Run time scenarios (macro)
...
+    Then Run the Pipeline in Runtime
+    Then Enter runtime argument value "databaseName" for key "DatabaseName"
+    Then Enter runtime argument value "fetchSize" for key "fetchSize"
+    And Run the Pipeline in Runtime with runtime arguments
+    Then Wait till pipeline is in running state
+    Then Open and capture logs
+    Then Verify the pipeline status is "Succeeded"
+
+    @MSSQL_SOURCE_TEST @MSSQL_SINK_TEST @Mssql_Required
+    Scenario: Verify pipeline failure message in logs when user provides invalid Table name in importQuery of plugin with MacrosScenario: To verify data is getting transferred from Mssql to Mssql successfully
```
Comment (Vipinofficial11) [inline mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTimeWithMacros.feature]: Same comment here we are not validating the failure message in logs. There is a step to verify logs message use that step in all the scenarios where you are validating error message.

File: mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTimeWithMacros.feature [hunk 11]
```diff
@@ -115,3 +115,169 @@ Feature: Mssql Source - Run time scenarios (macro)
...
+      Then Close the Plugin Properties page
+      And Save and Deploy Pipeline
+      And Run the Pipeline in Runtime
+      And Enter runtime argument value "invalidSelectQuery" for key "importQuery"
+      And Run the Pipeline in Runtime with runtime arguments
+      Then Wait till pipeline is in running state
+      And Verify the pipeline status is "Failed"
+
+      @MSSQL_SOURCE_TEST @MSSQL_SINK_TEST @Mssql_Required
+      Scenario: Verify pipeline failure message in logs when user provides invalid Credentials for connection with Macros
```
Comment (Vipinofficial11) [inline mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTimeWithMacros.feature:250]: Same comment here.

File: mssql-plugin/src/e2e-test/java/io.cdap.plugin/MssqlClient.java [hunk 12]
```diff
@@ -165,6 +165,16 @@ public static void deleteTables(String schema, String[] tables)
         }
     }
 
+    public static void createTargetMssqlTable(String targetTable, String schema) throws SQLException,
+            ClassNotFoundException {
+        try (Connection connect = getMssqlConnection();
+             Statement statement = connect.createStatement()) {
+            String datatypesColumns = PluginPropertyUtils.pluginProp("bigQueryDatatypesColumns");
```
Comment (Vipinofficial11) [inline mssql-plugin/src/e2e-test/java/io.cdap.plugin/MssqlClient.java]: Why are we using ```PluginPropertyUtils.pluginProp("bigQueryDatatypesColumns")``` is this creating columns in Big Query ?

File: mssql-plugin/src/e2e-test/resources/pluginParameters.properties [hunk 13]
```diff
@@ -42,7 +47,7 @@ InsertBQDataQueryFile=testData/BigQuery/BigQueryInsertDataQuery.txt
 
 datatypeColumns=(ID VARCHAR(100) PRIMARY KEY, COL1 TINYINT, COL2 SMALLINT, COL3 INT, COL4 BIGINT,\
   COL5 BIT, COL6 DECIMAL(38, 0), COL7 DECIMAL(10, 4), COL8 DECIMAL(5, 0), COL9 SMALLMONEY, COL10 MONEY, \
-  COL11 NUMERIC(38, 0), COL12 NUMERIC(10, 4), COL13 FLOAT, COL14 REAL, COL15 DATE, COL20 TIME(0), COL21 CHAR(100), \
+  COL11 NUMERIC(38, 0), COL12 NUMERIC(10, 4),bqOutputMultipleDatatypesSchema, COL14 REAL, COL15 DATE, COL20 TIME(0), COL21 CHAR(100), \
```
Comment (Vipinofficial11) [inline mssql-plugin/src/e2e-test/resources/pluginParameters.properties]: Revert this change you have altered one of the columns with this change.

File: mssql-plugin/src/e2e-test/java/io.cdap.plugin/MssqlClient.java [hunk 14]
```diff
@@ -169,7 +169,7 @@ public static void createTargetMssqlTable(String targetTable, String schema) thr
             ClassNotFoundException {
         try (Connection connect = getMssqlConnection();
              Statement statement = connect.createStatement()) {
-            String datatypesColumns = PluginPropertyUtils.pluginProp("SqlServerDatatypesColumns");
+            String datatypesColumns = PluginPropertyUtils.pluginProp("CloudMySqlDatatypesColumns");
```
Comment (Vipinofficial11) [inline mssql-plugin/src/e2e-test/java/io.cdap.plugin/MssqlClient.java]: This is SQL Server why name is CloudMySQL

File: mssql-plugin/src/e2e-test/features/mssql/mssql sink/RunTimeWithMacros.feature (hunk 8 above)
Comment (priyabhatnagar25) [reply mssql-plugin/src/e2e-test/features/mssql/mssql sink/RunTimeWithMacros.feature:117]: resolved

File: mssql-plugin/src/e2e-test/features/mssql/mssql sink/RunTimeWithMacros.feature (hunk 9 above)
Comment (priyabhatnagar25) [reply mssql-plugin/src/e2e-test/features/mssql/mssql sink/RunTimeWithMacros.feature:160]: resolved

File: mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTimeWithMacros.feature (hunk 10 above)
Comment (priyabhatnagar25) [reply mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTimeWithMacros.feature]: resolved

File: mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTimeWithMacros.feature (hunk 11 above)
Comment (priyabhatnagar25) [reply mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTimeWithMacros.feature:250]: resolved

File: mssql-plugin/src/e2e-test/resources/pluginParameters.properties (hunk 13 above)
Comment (priyabhatnagar25) [reply mssql-plugin/src/e2e-test/resources/pluginParameters.properties]: resolved

File: mssql-plugin/src/e2e-test/resources/pluginParameters.properties (hunk 7 above)
Comment (priyabhatnagar25) [reply mssql-plugin/src/e2e-test/resources/pluginParameters.properties]: resolved

File: mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature [hunk 15]
```diff
@@ -147,40 +147,3 @@ Feature: Mssql - Verify Mssql source data transfer
     Then Verify the pipeline status is "Succeeded"
     Then Close the pipeline logs
     Then Validate the values of records transferred to target BigQuery table is equal to the values from source Table
-
-  @MSSQL_SOURCE_TEST @BQ_SINK @Mssql_Required
-  Scenario: To verify the pipeline fails while preview with invalid bounding query setting the required split-By field
```
Comment (Vipinofficial11) [inline mssql-plugin/src/e2e-test/features/mssql/mssql source/RunTime.feature]: @priyabhatnagar25  Why have you deleted this test ?

**Comment by Vipinofficial11** [comment]
To do : Cherry pick (df88dac15fcbab5abf42bc43825ff9c009b3ad5a) in external PR and Close this Internal PR.
